proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,634 @@
+
+.. [#] See `Do we need different syntax for "local" imports?`_ for discussions
+   of the consequences.
+
+Module aliases
+~~~~~~~~~~~~~~
+
+We propose adding a language extension ``ModuleAliases``. This allows a
+new kind of top-level definition: ``module <name> = <qualname>``
+
+A module alias adds a binding in the enclosing scope from the new name
+to the module referred to by the right hand side.
+
+Examples
+--------
+
+Exporting small utility modules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+::
+
+   module A (f, module Unsafe) where
+
+   f = Unsafe.g …
+
+   module Unsafe where
+       g = … 
+
+Exporting qualified names
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This proposal allows exporting modules, but doesn’t say they can be
+exported under *qualified* names, which is something that people want to
+do. Consider the following `proposed
+example <https://github.com/ghc-proposals/ghc-proposals/pull/283#issuecomment-541525245>`_:
+
+::
+
+   module MyPrelude ( qualified module BL
+                    , qualified module BS
+                    , Set
+                    , qualified module Set ) where
+
+
+   import qualified Data.ByteString.Lazy as BL
+   import qualified Data.ByteString as BS
+   import Data.Set ( Set )
+   import qualified Data.Set as Set
+
+This could be achieved as follows:
+
+::
+
+   module MyPrelude (module Data, Set) where
+
+   import qualified Data.ByteString.Lazy as BL
+   import qualified Data.ByteString as BS
+   import Data.Set ( Set )
+   import qualified Data.Set as Set
+
+
+   module Data (module ByteString, module Set) where
+       module ByteString (module Lazy, module BS (..)) where
+           import BS
+           module Lazy (module BSL (..)) where
+               import BSL
+       module Set (module Set (..)) where
+           import Set
+
+Because qualified names correspond to module-qualification, we can
+always create that structure ourselves. This is somewhat laborious for
+the author, but does allow the desired experience for the consumer.
+
+This example is somewhat unsatisfactory, since one would really like to
+re-export the “module” ``Data.ByteString``, but this is not a module as
+far as this proposal is concerned. See `Can we make dotted names less
+special?`_ for more discussion.
+
+Effect and interactions
+-----------------------
+
+I believe this proposal would solve all the problems described in [ghc283]_ and [ghc205]_.
+Moreover, because it makes modules just like other entities, improvements to the module
+system will equally benefit management of modules.
+
+Interactions
+~~~~~~~~~~~~
+
+`Modules as entities`_ should have no interaction with any other
+features. We do not even need to change interface files, since module
+names cannot be exported.
+
+`Module export specifiers`_ has slightly more effects:
+
+- Interface files must be able to handle the possibility that an exported name refers to a
+  module. This may have some interaction with Backpack.
```

Comment:
There's the question of the serialisation -- should it be:

1. A reference to a module (that'll have consequences that are difficult to predict, I'm afraid), or
2. A flat list of names-qualified-with-dotted-names -- which appears to be much more limited in impact -- as per https://github.com/ghc-proposals/ghc-proposals/pull/295/files#r347111957

Reply:
I actually think the backpack situation will be too bad. Modules can already reexport siganture's items, for example. If I had to make a guess, the export will be restricted to be a module interface with the items in the original signatures, but the types of those items will be richer due to later instantiations.

Reply:
I don't know that much about how interface files currently work. However, here's a sketch of a way to structure them that would work fine for exporting modules, even recursive ones.

```
type Scope = Map Name Symbol
type SymbolTable = Map Symbol Entity
data Entity = Type ... | Value ... | Module Scope
-- The `Scope` is the exported scope of the module
data Interface = Interface Scope SymbolTable
```